### DISCUSSION TITLE
[@types/json-schema]: reasoning behind `| boolean` for main schema object type definition

### DISCUSSION THREAD

**Discussion by camnicklaus** [DefinitelyTyped/DefinitelyTyped#55043] [_Misc]
reference: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/ee521ca782c6d9d3d801f0bf5b301b225e3e3842/types/json-schema/index.d.ts#L617

I'm fairly newish to typescript and am working to convert some utilities that process/traverse json-schema. 
I'm finding that, when attempting to access nested properties, I'm having to write conditions like:
`if (typeof jsonSchema?.properties?.myProperty === 'boolean') return;`

all over the place to avoid typescript errors and it just doesn't feel right, like maybe I could be setting something up differently somehow?

apologies, this feels much more like a general typescript question than an issue with the json-schema types in this repo but this is the first context where I've run into this.

**Comment by KapitanOczywisty** [comment]
Schema in some cases can be replaced with `true` or `false` and typings are reflecting that. For me these types are useful for writing schemas in TS, for reading you may have better experiences using [`json-pointer`](https://www.npmjs.com/package/json-pointer) or something like that.

**Reply by camnicklaus** [reply]
Thanks so much for the reply! 
I do think a lib for traversing may be really helpful but wouldn't I still have the same issue?
In other words, even if I used json-pointer's `get` method, the returned node (apologies if that's an incorrect term) would have the same type definition, so if I used it to access a nested property,  since the type of the value for `properties` is `JSONSchema7Definition | undefined`: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/ee521ca782c6d9d3d801f0bf5b301b225e3e3842/types/json-schema/index.d.ts#L663, I'd still need to guard the value like so: `typeof value === 'boolean'` no?

I might just be suffering from my own typescript knowledge gaps, but it seems like, since `{ properties: true }` would not be valid json-schema we'd always have to guard in order to avoid typescript errors? 
again, apologies if I'm missing something super obvious here 😬

**Reply by KapitanOczywisty** [reply]
All depends on what are you doing. `get` returns `any`, so for known schama you will probably do `const item: JSONSchema7 = pointer.get('/properties/myProperty');`. In other cases you should probably check for boolean just in case. For traversing `walk` or `dict` might be useful.

You can make own type guards to make boolean check a bit cleaner:
```ts
function isJSONSchema7(schema: JSONSchema7Definition): schema is JSONSchema7 {
  return schema && typeof schema !== 'boolean';
}

if(isJSONSchema7(jsonSchema?.properties?.myProperty)){
  // schema
} else {
  // boolean
}

```

**Reply by camnicklaus** [reply]
Ah, I think that makes sense. thanks so much for bearing with me ❤️ 🍻
